core/src/main/java/org/sonarsource/sonarlint/core/SonarLintClientImpl.java

```diff
@@ -45,54 +50,216 @@
-    checkStarted();
+    checkNotNull(issueListener);
+    rwl.readLock().lock();
+    try {
+      checkStarted();
+      try {
+        return globalContainer.analyze(configuration, issueListener);
+      } catch (RuntimeException e) {
+        throw handleException(e);
+      }
+    } finally {
+      rwl.readLock().unlock();
+    }
+  }
+
+  @Override
+  public GlobalSyncStatus getSyncStatus() {
+    checkMode();
+    rwl.readLock().lock();
+    try {
+      checkStarted();
+      return ((StorageGlobalContainer) globalContainer).getSyncStatus();
+    } finally {
+      rwl.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void sync(ServerConfiguration serverConfig) {
+    checkNotNull(serverConfig);
+    checkMode();
+    rwl.writeLock().lock();
+    try {
+      checkStopped();
+      ConnectedContainer connectedContainer = new ConnectedContainer(globalConfig, serverConfig);
+      try {
+        connectedContainer.startComponents();
+      } catch (RuntimeException e) {
+        throw handleException(e);
+      }
+      try {
+        connectedContainer.sync();
+      } finally {
+        try {
+          connectedContainer.stopComponents(false);
+        } catch (RuntimeException e) {
+          throw handleException(e);
+        }
+      }
+    } finally {
+      rwl.writeLock().unlock();
+    }
+  }
+
+  private void checkMode() {
+    if (globalConfig.getServerId() == null) {
+      throw new UnsupportedOperationException("Unable to sync in unconnected mode");
+    }
+  }
+
+  @Override
+  public ValidationResult validateCredentials(ServerConfiguration serverConfig) {
+    checkNotNull(serverConfig);
+    checkMode();
+    rwl.readLock().lock();
+    try {
+      ConnectedContainer connectedContainer = new ConnectedContainer(globalConfig, serverConfig);
+      try {
+        connectedContainer.startComponents();
+      } catch (RuntimeException e) {
+        throw handleException(e);
+      }
+      try {
+        return connectedContainer.validateCredentials();
+      } finally {
+        try {
+          connectedContainer.stopComponents(false);
+        } catch (RuntimeException e) {
+          throw handleException(e);
+        }
+      }
+    } finally {
+      rwl.readLock().unlock();
+    }
+  }
+
+  @Override
+  public List<RemoteModule> searchModule(ServerConfiguration serverConfig, String exactKeyOrPartialName) {
+    checkNotNull(serverConfig);
+    checkMode();
+    rwl.readLock().lock();
+    try {
+      ConnectedContainer connectedContainer = new ConnectedContainer(globalConfig, serverConfig);
+      try {
+        connectedContainer.startComponents();
+      } catch (RuntimeException e) {
+        throw handleException(e);
+      }
+      try {
+        return connectedContainer.searchModule(exactKeyOrPartialName);
+      } finally {
+        try {
+          connectedContainer.stopComponents(false);
+        } catch (RuntimeException e) {
+          throw handleException(e);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Remove this throw statement from this finally block. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1143)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not throw exceptions in finally blocks. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1163)